psd-web/spec/support/elasticsearch.rb

```
@@ -1,7 +1,21 @@
 RSpec.shared_context "with stubbed Elasticsearch", shared_context: :metadata do
   before do
     elasticsearch_url = ENV.fetch("ELASTICSEARCH_URL", "http://elasticsearch:9200")
-    stub_request(:any, /#{Regexp.quote(elasticsearch_url)}/).to_return(body: "Elasticsearch disabled in Rspec", status: 200)
+    empty_elasticsearch_response = {
```

This might not be intuitive...

---

That's an actual ES response. I had some really hard time debugging the "Elasticsearch disabled in Rspec" which would not appear anywhere has it's not JSON parseable and I i'd get some really funky behaviour.
I'd suggest we keep it that way it's `\/cases||\/products||\/businesses` expect to this shapre to work properly with the `ElasticSearch::ResultProxy`.

I'm open to suggestion though

---

I think we need some indication that it is a stubbed response and not just a real Elasticsearch producing no results. Perhaps a log message (that could be too noisy) or a fake result which is obviously a result of the stub?

---

I'd have expected the `:with_stubbed_elasticsearch` metadata to be clear enough. 
A "fake" result would probably have some weird side effect, also I'm not sure what you mean by "fake" result. Agreed that the logging will be too noisy. 

I'll try to remove it all together but that might re-throw a 400 error. I spent an entire day trying to figure out what was going on.

---

```
     ActionView::Template::Error:
       undefined method `each_pair' for "Elasticsearch disabled in Rspec":String
       Did you mean?  each_char
     # /Users/yann/.gem/ruby/2.5.7/gems/hashie-3.6.0/lib/hashie/mash.rb:210:in `deep_update'
     # /Users/yann/.gem/ruby/2.5.7/gems/hashie-3.6.0/lib/hashie/mash.rb:124:in `initialize'
     # /Users/yann/.gem/ruby/2.5.7/gems/elasticsearch-model-6.0.0/lib/elasticsearch/model/response.rb:30:in `new'
     # /Users/yann/.gem/ruby/2.5.7/gems/elasticsearch-model-6.0.0/lib/elasticsearch/model/response.rb:30:in `response'
     # /Users/yann/.gem/ruby/2.5.7/gems/elasticsearch-model-6.0.0/lib/elasticsearch/model/response/records.rb:37:in `ids'
     # /Users/yann/.gem/ruby/2.5.7/gems/elasticsearch-model-6.0.0/lib/elasticsearch/model/adapters/active_record.rb:22:in `records'
     # /Users/yann/.gem/ruby/2.5.7/gems/elasticsearch-model-6.0.0/lib/elasticsearch/model/response/records.rb:13:in `each'
     # /Users/yann/.gem/ruby/2.5.7/gems/draper-3.1.0/lib/draper/collection_decorator.rb:43:in `map'
     # /Users/yann/.gem/ruby/2.5.7/gems/draper-3.1.0/lib/draper/collection_decorator.rb:43:in `decorated_collection'
     # /Users/yann/.gem/ruby/2.5.7/gems/draper-3.1.0/lib/draper/delegation.rb:10:in `any?'
     # ./app/views/investigations/index.html.slim:9:in `_app_views_investigations_index_html_slim___1924593218742029828_70276346109640'
     # /Users/yann/.gem/ruby/2.5.7/gems/sentry-raven-2.13.0/lib/raven/integrations/rails/controller_transaction.rb:7:in `block in included'
     # /Users/yann/.gem/ruby/2.5.7/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:420:in `call_app!'
     # /Users/yann/.gem/ruby/2.5.7/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:274:in `mock_call!'
     # /Users/yann/.gem/ruby/2.5.7/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:186:in `call!'
     # /Users/yann/.gem/ruby/2.5.7/gems/omniauth-1.9.0/lib/omniauth/strategy.rb:169:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/scout_apm-2.6.6/lib/scout_apm/middleware.rb:20:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/warden-1.2.8/lib/warden/manager.rb:36:in `block in call'
     # /Users/yann/.gem/ruby/2.5.7/gems/warden-1.2.8/lib/warden/manager.rb:34:in `catch'
     # /Users/yann/.gem/ruby/2.5.7/gems/warden-1.2.8/lib/warden/manager.rb:34:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/tempfile_reaper.rb:15:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/etag.rb:27:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/conditional_get.rb:27:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/head.rb:12:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/session/abstract/id.rb:265:in `context'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/session/abstract/id.rb:259:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/railties-5.2.3/lib/rails/rack/logger.rb:38:in `call_app'
     # /Users/yann/.gem/ruby/2.5.7/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `block in call'
     # /Users/yann/.gem/ruby/2.5.7/gems/railties-5.2.3/lib/rails/rack/logger.rb:26:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/request_store-1.5.0/lib/request_store/middleware.rb:19:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/method_override.rb:24:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/runtime.rb:22:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/sendfile.rb:110:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/sentry-raven-2.13.0/lib/raven/integrations/rack.rb:51:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/railties-5.2.3/lib/rails/engine.rb:524:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/urlmap.rb:74:in `block in call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/urlmap.rb:58:in `each'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-2.2.1/lib/rack/urlmap.rb:58:in `call'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-test-1.1.0/lib/rack/mock_session.rb:29:in `request'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-test-1.1.0/lib/rack/test.rb:266:in `process_request'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-test-1.1.0/lib/rack/test.rb:129:in `custom_request'
     # /Users/yann/.gem/ruby/2.5.7/gems/rack-test-1.1.0/lib/rack/test.rb:58:in `get'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/browser.rb:68:in `process'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/browser.rb:52:in `block in process_and_follow_redirects'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/browser.rb:47:in `times'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/browser.rb:47:in `process_and_follow_redirects'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/browser.rb:23:in `visit'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/rack_test/driver.rb:45:in `visit'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/session.rb:278:in `visit'
     # /Users/yann/.gem/ruby/2.5.7/gems/capybara-3.31.0/lib/capybara/dsl.rb:58:in `block (2 levels) in <module:DSL>'
     # ./spec/support/login_helpers.rb:21:in `sign_in'
     # ./spec/features/send_alert_spec.rb:10:in `block (2 levels) in <top (required)>'
     # ./spec/support/keycloak.rb:14:in `block (2 levels) in <top (required)>'
     # ./spec/support/database_cleaner.rb:9:in `block (3 levels) in <top (required)>'
     # /Users/yann/.gem/ruby/2.5.7/gems/database_cleaner-1.8.2/lib/database_cleaner/generic/base.rb:16:in `cleaning'
     # /Users/yann/.gem/ruby/2.5.7/gems/database_cleaner-1.8.2/lib/database_cleaner/configuration.rb:87:in `block (2 levels) in cleaning'
     # /Users/yann/.gem/ruby/2.5.7/gems/database_cleaner-1.8.2/lib/database_cleaner/configuration.rb:88:in `cleaning'
     # ./spec/support/database_cleaner.rb:8:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # NoMethodError:
     #   undefined method `each_pair' for "Elasticsearch disabled in Rspec":String
     #   Did you mean?  each_char
     #   /Users/yann/.gem/ruby/2.5.7/gems/hashie-3.6.0/lib/hashie/mash.rb:210:in `deep_update'
```

---

@slorek that's error i get ^^ It's not fun to debug.

---

`undefined method `each_pair' for "Elasticsearch disabled in Rspec":String` is quite obvious IMO or you could raise an exception, but I'm not too fussed as long as there's some easy way to see that it's stubbed. Otherwise it could be a big time sink for someone in future 😂